### PR TITLE
Add domain logic for push notifications

### PR DESCRIPTION
**Pull request by iamacook**
## Summary

This adds the relative logic to the domain for enqueuing notifications to subscribers when the relevant hook event is received.

It leverages the `NotificationsDatasource` to `getSubscribersWithTokensBySafe` for the relative `TransactionEventType` and `enqueueNotification`s accordingly.

## Changes

- Add `INotificationsRepositoryV2` behind a feature flag:
  - `enqueueNotification`
  - `upsertSubscriptions`
  - `getSafeSubscription`
  - `getSubscribersWithTokensBySafe`
  - `deleteSubscription`
  - `deleteDevice`
- Extend `HooksRepository` with `onEventEnqueueNotifications` with respective entities.
- Add `transaction_hash` query to `getIncomingTransfers` and propagate changes/add tests.
- Locate previously domain-relative logic for notifications in domain, e.g. entities and builders.
- Add approriate test coverage.

### CODE REVIEW

File: src/domain/notifications/notifications.repository.v2.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+          chainId: safeToSubscribe.chainId,
+          safeAddress: safeToSubscribe.address,
+          delegate: args.account,
+        })
+        .catch(() => null);
+
+      // Upsert delegate
+      if (
+        delegates &&
+        delegates.results.some(({ delegate }) => delegate === args.account)
```
Comment (iamacook) [inline src/domain/notifications/notifications.repository.v2.ts:63]: These checks need to be disucssed as it is still unclear whether we allow anyone to subscribe to queue-related events.

File: src/domain/notifications/notifications.repository.v2.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        .catch(() => null);
+
+      const isDelegate = !!delegates?.results.some(
+        ({ delegate }) => delegate === args.account,
+      );
+      if (isDelegate) {
+        continue;
+      }
+
+      throw new UnauthorizedException();
```
Comment (hectorgomezv) [inline src/domain/notifications/notifications.repository.v2.ts:123]: Nit: should we log this error (the account signer not being a owner/delegate of one of the target Safes), wdyt?

**Review by hectorgomezv — APPROVED** [review]

File: src/routes/hooks/hooks-notifications.spec.ts [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+    notificationsDatasource.getSubscribersBySafe.mockResolvedValue(subscribers);
+    const message = messageBuilder()
+      .with('messageHash', event.messageHash as `0x${string}`)
+      .with(
+        'confirmations',
+        subscribers.map((subscriber) => {
+          return messageConfirmationBuilder()
+            .with('owner', subscriber.subscriber)
+            .build();
+        }),
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/routes/hooks/hooks-notifications.spec.ts:727]: Added in 663d9016dd7226f0da55c1cef1048f4fa8ef02d3.

File: src/domain/notifications/notifications.repository.v2.ts (hunk 2 above)
Comment (iamacook) [reply src/domain/notifications/notifications.repository.v2.ts:125]: Added in 663d9016dd7226f0da55c1cef1048f4fa8ef02d3.

**Review by hectorgomezv — APPROVED** [review]

**Comment by hectorgomezv** [comment]
I think this PR needs a rebase onto the source branch 🙂

**Review by hectorgomezv — APPROVED** [review]

**Review by hectorgomezv — APPROVED** [review]

**Pull request merged**